p-ata/src/processor.rs

```diff
@@ -0,0 +1,982 @@
+        })
+    }
+}
+
+/// Check if account already exists and is properly configured (idempotent check).
+///
+/// This function validates that an existing ATA account:
+/// 1. Is owned by the correct token program
+/// 2. Has the correct owner (wallet)
+/// 3. Has the correct mint
+/// 4. Uses the canonical PDA address, even when `expected_bump` is provided
+///
+/// ## Returns
+///
+/// * `Ok(true)` - Account exists and is properly configured (safe to skip creation)
+/// * `Ok(false)` - Account doesn't exist or is not token program owned (needs creation)  
+/// * `Err(_)` - Account exists but has invalid configuration (error condition)
+#[inline(always)]
+pub(crate) fn check_idempotent_account(
+    associated_token_account: &AccountInfo,
+    wallet: &AccountInfo,
+    mint_account: &AccountInfo,
+    token_program: &AccountInfo,
+    idempotent: bool,
+    program_id: &Pubkey,
+    expected_bump: Option<u8>,
+) -> Result<bool, ProgramError> {
+    if idempotent && associated_token_account.is_owned_by(token_program.key()) {
+        let ata_state = get_token_account(associated_token_account)?;
+
+        validate_token_account_owner(ata_state, wallet.key())?;
+        validate_token_account_mint(ata_state, mint_account.key())?;
+
+        match expected_bump {
+            Some(bump) => {
+                let seeds: &[&[u8]; 3] = &[
+                    wallet.key().as_ref(),
+                    token_program.key().as_ref(),
+                    mint_account.key().as_ref(),
+                ];
+
+                // Check if a better canonical bump exists
+                reject_if_better_valid_bump_exists(seeds, program_id, bump)?;
+
+                let maybe_canonical_address = derive_address::<3>(seeds, Some(bump), program_id);
+
+                // We must check that the actual derived address is off-curve,
+                // since it will not fail downstream as in Create paths.
+                // Potential problem if skipping this is demonstrated in
+                // tests/bump/test_idemp_oncurve_attack.rs
+                if !is_off_curve(&maybe_canonical_address)
+                    || maybe_canonical_address != *associated_token_account.key()
+                {
+                    log!(
+                        "Error: Provided `expected_bump` {} is on curve and non-canonical.",
+                        bump
+                    );
+                    return Err(ProgramError::InvalidSeeds);
+                }
+            }
+            None => {
+                let (canonical_address, _bump) = derive_canonical_ata_pda(
+                    wallet.key(),
+                    token_program.key(),
+                    mint_account.key(),
+                    program_id,
+                );
+
+                if canonical_address != *associated_token_account.key() {
+                    log!(
+                        "Error: Provided associated token account address is not canonical. Expected: {}, Found: {}",
```

Comment:
Same here about the "non-canonical" message – I think this is a mismatch instead of non-canonical.